### PR TITLE
Fixed a bug with settling IOU

### PR DESCRIPTION
**Pull request by ischasny**
Fixed a couple of bugs with settling an IOU:
1. Anonymous identities haven't been synced between the parties
2. The borrower didn't take into account anonymous keys when signed
3. The lender always expected an IOUState to be in Tx outputs, which failed when IOU is paid off

### CODE REVIEW

**Review by kupham — APPROVED** [review]

File: src/main/kotlin/net/corda/training/flow/IOUSettleFlow.kt [hunk 1]
```diff
@@ -45,16 +46,15 @@ class IOUSettleFlow(val linearId: UniqueIdentifier, val amount: Amount<Currency>
 
         // Step 4. Check we have enough cash to settle the requested amount.
         val cashBalance = serviceHub.getCashBalance(amount.token)
-        if (cashBalance == null) {
-            throw IllegalArgumentException("Borrower has no ${amount.token} to settle.")
```
Comment (amoothart) [inline src/main/kotlin/net/corda/training/flow/IOUSettleFlow.kt:49]: There is a unit test which depends on this specific  message which will now likely fail: borrowerMustHaveCashInRightCurrency

File: src/main/kotlin/net/corda/training/flow/IOUSettleFlow.kt [hunk 2]
```diff
@@ -71,11 +71,17 @@ class IOUSettleFlow(val linearId: UniqueIdentifier, val amount: Amount<Currency>
 
         // Step 8. Verify and sign the transaction.
         builder.verify(serviceHub)
-        val ptx = serviceHub.signInitialTransaction(builder)
+        val myKeysToSign = cashKeys + ourIdentity.owningKey
```
Comment (amoothart) [inline src/main/kotlin/net/corda/training/flow/IOUSettleFlow.kt]: Can you add some comments either here or in the unit tests on what the cashKeys are and how they get signed. IMO this is one of the more confusing parts of flows and we should document thoroughly.

File: src/main/kotlin/net/corda/training/flow/IOUSettleFlow.kt [hunk 3]
```diff
@@ -90,10 +96,13 @@ class IOUSettleFlow(val linearId: UniqueIdentifier, val amount: Amount<Currency>
...
     override fun call() {
+
+        // Receiving information about anonymous identities
+        subFlow(IdentitySyncFlow.Receive(flowSession))
+
+        // signing transaction
         val signedTransactionFlow = object : SignTransactionFlow(flowSession) {
-            override fun checkTransaction(stx: SignedTransaction) = requireThat {
-                val outputStates = stx.tx.outputs.map { it.data::class.java.name }.toList()
-                "There must be an IOU transaction." using (outputStates.contains(IOUState::class.java.name))
```
Comment (amoothart) [inline src/main/kotlin/net/corda/training/flow/IOUSettleFlow.kt:96]: Why did you remove the responder check?

**Review by amoothart — COMMENTED** [review]
A couple thoughts

File: src/main/kotlin/net/corda/training/flow/IOUSettleFlow.kt (hunk 1 above)
Comment (ischasny) [reply src/main/kotlin/net/corda/training/flow/IOUSettleFlow.kt:49]: It can't be null as cashBalance is not nullable. So null check was redundant.

File: src/main/kotlin/net/corda/training/flow/IOUSettleFlow.kt (hunk 2 above)
Comment (ischasny) [reply src/main/kotlin/net/corda/training/flow/IOUSettleFlow.kt]: Yeah, this is a bit which is still missing in documentation. This is also something, we should put an emphasis on during the trainings as anonymous parties is one of great features in Corda. Added more notes there.

File: src/main/kotlin/net/corda/training/flow/IOUSettleFlow.kt (hunk 3 above)
Comment (ischasny) [reply src/main/kotlin/net/corda/training/flow/IOUSettleFlow.kt:96]: As it fails when IOU is paid off and retired from the ledger. It contradicts the contract code if you check.

File: src/main/kotlin/net/corda/training/flow/IOUSettleFlow.kt (hunk 1 above)
Comment (amoothart) [reply src/main/kotlin/net/corda/training/flow/IOUSettleFlow.kt:49]: That is true but make sure the removal doesn't affect the associated test: https://github.com/roger3cev/corda-training-solutions/blob/master/src/test/kotlin/net/corda/training/flow/IOUSettleFlowTests.kt#L158

Comment (ischasny) [reply src/main/kotlin/net/corda/training/flow/IOUSettleFlow.kt:49]: Yeps, all looks good.

**Comment by amoothart** [comment]
LGTM
